plugins/woocommerce/src/Admin/API/AI/Product.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace Automattic\WooCommerce\Admin\API\AI;
+
+use Automattic\WooCommerce\Blocks\AIContent\UpdateProducts;
+use WP_Error;
+use WP_REST_Request;
+use WP_REST_Response;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Product controller
+ *
+ * @internal
+ */
+class Product {
+
+	/**
+	 * Endpoint namespace.
+	 *
+	 * @var string
+	 */
+	protected $namespace = 'wc-admin';
+
+	/**
+	 * Route base.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'ai';
+
+	/**
+	 * Register routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/product',
+			array(
+				array(
+					'methods'             => \WP_REST_Server::CREATABLE,
+					'callback'            => array( $this, 'update_product' ),
+					'permission_callback' => array( Middleware::class, 'is_authorized' ),
+					'args'                => array(
+						'products_information' => array(
+							'description' => __( 'Data generated by AI for updating dummy products.', 'woocommerce' ),
+							'type'        => 'object',
+						),
+						'last_product'         => array(
+							'description' => __( 'Whether the product being updated is the last one in the loop', 'woocommerce' ),
+							'type'        => 'boolean',
+						),
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Update product with the content and image powered by AI.
+	 *
+	 * @param  WP_REST_Request $request Request object.
+	 *
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function update_product( WP_REST_Request $request ) {
+		$product_updater     = new UpdateProducts();
+		$product_information = $request['products_information'] ?? array();
+
+		if ( empty( $product_information ) ) {
+			return rest_ensure_response(
+				array(
+					'ai_content_generated' => true,
+				)
+			);
+		}
+
+		$product_updater->update_product_content( $product_information );
+
+		$last_product_to_update = $request['last_product'] ?? false;
+
+		if ( $last_product_to_update ) {
+			flush_rewrite_rules();
```

Comment:
This is an expensive operation, the [documentation](https://developer.wordpress.org/reference/functions/flush_rewrite_rules/) even explicitly states as much.

> However, this is an expensive operation so it should only be used when necessary.

I am wondering if we can optimize this further somehow to avoid executing it more times than is necessary, if you think it has potential to do so? For example, we could use a transient to limit how often this occurs.

Reply:
When we implemented this we were flushing after every execution request to the endpoint and we did this optimization https://github.com/woocommerce/woocommerce-blocks/pull/11971 to only do it after the last.
Tbh, I wouldn't invest much more time in further optimizing this flow since it's not going to be maintained for much longer 🤔